lib/ide/parser/src/api.rs

```diff
@@ -3,9 +3,57 @@
 use crate::prelude::*;
 
 use ast::IdMap;
+use ast::HasRepr;
+use ast::HasIdMap;
 
 pub use ast::Ast;
 
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use serde::Serialize;
+
+
+
+// ================
+// == SourceFile ==
+// ================
+
+/// Things that are metadata.
+pub trait Metadata:Serialize+DeserializeOwned {}
+
+/// Raw netadata.
+impl Metadata for serde_json::Value {}
+
+/// Parsed file / module with metadata.
+#[derive(Debug,Clone,Serialize,Deserialize,PartialEq,Eq)]
+pub struct SourceFile<M:Metadata> {
+    /// Ast representation.
+    pub ast: Ast,
+    /// Raw metadata in json.
+    #[serde(bound(deserialize = ""))]
+    pub metadata: M
+}
```

Comment:
Oh here is one more important thing. Never ever make constraints in struct definitions. This is bad. It was banned in haskell for a reason. Just dont do it. Use bounds on impls